### PR TITLE
Settings Command Spec

### PR DESCRIPTION
**Pull request by msftrubengu**
Spec for #396 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/winget-cli/pull/402)

### CODE REVIEW

File: doc/specs/#396 - Settings command.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+### Accessibility
+
+Any description or output text that is added by consequence of this feature will need to be localized. Moreover, this feature allows a mechanism to add accessibility settings in the future.
+
+### Security
+
+This should not introduce any _new_ security concerns.
+
+### Reliability
```
Comment (JohnMcPMS) [inline doc/specs/#396 - Settings command.md:150]: Consider here, or somewhere, describing what will happen if the file can be parsed, but some value is outside of its defined range.

Will we ignore individual settings?  Or will we treat it as an error all up and load the backup?

File: doc/specs/#396 - Settings command.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+If the file isn't intuitive users will experience difficulties setting what they need.
+
+## Future considerations
+
+This feature allows the ability to expand the customization of winget for any user. It is also design to be expand with future settings commands that doesn't fit the scope of this future. For example, one could see `winget settings set SOME_SETTING on` to modify the settings without the need of editing a file.
+
+## References
+
+@JohnMcPMS for telling my what to type.
```
Comment (JohnMcPMS) [inline doc/specs/#396 - Settings command.md:174]: I just tell you what I think you should type.

File: doc/specs/#396 - Settings command.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+### Format
+
+The WinGet settings file needs to be in a readable format for users.
+
+#### Option 1: YAML 
+
+WinGet already knows how to handle YAML files via yaml-cpp. To follow the manifest style, the properties will be PascalCased.
+
+#### Option 2: JSON
```
Comment (JohnMcPMS) [inline doc/specs/#396 - Settings command.md:30]: I think JSON is the right choice, as it is more natural here.  But we do need to determine how will yaml-cpp parses JSON with comments.

Comment (yao-msft) [reply doc/specs/#396 - Settings command.md:30]: Shall we use jsoncpp from the terminal PR that Ruben mentioned previously?

File: doc/specs/#396 - Settings command.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+A command is a better option than having an argument, for example `winget --settings`, because it let us add more commands into it in the future such as set and unset.
+
+### Documentation
+
+All setting must be documented in the winget-cli repository in doc\Settings.md. The settings file will need to have a link to this file or some other Microsoft documentation site for reference. However, having a comment on JSON wouldn't be ideal.
+
+### Backup Settings File
+
+A user might make a mistake could make the settings file unparsable. To protect against this there must be a backup settings file with the latest known good settings file. The flow is the following:
```
[the remark this answers is not in the document]
Comment (msftrubengu) [reply doc/specs/#396 - Settings command.md:66]: A lot more work than just ShellExecute

File: doc/specs/#396 - Settings command.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+1. winget settings
+2. If settings file can be parsed copy to settings.backup
```
[the remark this answers is not in the document]
Comment (msftrubengu) [reply doc/specs/#396 - Settings command.md:69]: I added more information in the loading settings section

File: doc/specs/#396 - Settings command.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+### Compatibility
+
+Previous versions winget will not be able to read a setting file. An update to winget is required.
+
+### Performance, Power, and Efficiency
+
+This introduce a read file for every command being run. However this shouldn't affect the performance of WinGet.
+
+## Potential Issues
```
[the remark this answers is not in the document]
Comment (msftrubengu) [reply doc/specs/#396 - Settings command.md:202]: Clean up will only be done when winget runs in a package context. When it doesn't there's no really a good way to clean the file

File: doc/specs/#396 - Settings command.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+## Capabilities
+
+It will allow winget to set and define any new settings in the future.
+
+### Accessibility
+
+Any description or output text that is added by consequence of this feature will need to be localized. Moreover, this feature allows a mechanism to add accessibility settings in the future.
+
+### Security
```
[the remark this answers is not in the document]
Comment (msftrubengu) [reply doc/specs/#396 - Settings command.md:184]: I'll investigate more on this

File: doc/specs/#396 - Settings command.md (hunk 3 above)
Comment (riverar) [reply doc/specs/#396 - Settings command.md:30]: _Another_ file format? Pick one, don't mix and match. That's confusing.

File: doc/specs/#396 - Settings command.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+#### Option 2: JSON
+
+JSON is the popular formats that most applications tend to use. It is also possible to parse JSON with yaml-cpp as YAML is a superset of JSON, but a proper JSON parser might be needed depending on how well yaml-cpp handles it. Options are third party json parsers or Microsoft cpp winrt implementation `winrt::Windows::Data::Json::JsonObject`. Properties will be camelCased. An schema will need to be created as well.
+
+### Location
+
+WinGet can either run in package context or not. That means the location of the settings file will be determine depending on the context. 
+
+Package Context: %LOCALAPPDATA%\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\settings.yaml
```
[the remark this answers is not in the document]
Comment (riverar) [reply doc/specs/#396 - Settings command.md]: Would alternatively or additionally look for `%USERPROFILE%\.winget`

File: doc/specs/#396 - Settings command.md (hunk 3 above)
Comment (JohnMcPMS) [reply doc/specs/#396 - Settings command.md:30]: @riverar, is that a vote for YAML, or a vote against a JSON with non "standard" comments.  I'm interpreting it as for YAML to be consistent with the manifests, but making sure.

File: doc/specs/#396 - Settings command.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (JohnMcPMS) [reply doc/specs/#396 - Settings command.md]: For reference, Terminal stores their settings in exactly the same location (that is, the Local state location for their package).

File: doc/specs/#396 - Settings command.md (hunk 3 above)
Comment (riverar) [reply doc/specs/#396 - Settings command.md:30]: I'm voting for consistency with repository yaml just to keep the code base smaller, less risk of failure. Comments aren't important in my mind, I think docs/schema/ui can easily fill this gap.

**Review by Jokerok50 — COMMENTED** [review]
gl

**Pull request merged**